tips/tip-0002/tip-0002.md

```diff
@@ -0,0 +1,218 @@
+
+The formal indeterminate $X$ here anticipates the verifier's second challenge $\beta$. The prover computes $\mathsf{rp}_T$ symbolically before $\beta$ is supplied – in fact, this symbolic terminal is the polynomial $h(X)$.
+
+### Square-Freeness Argument
+
+All memory-like tables' clock jump differences are members of the Processor Table's `clk` column if $h(X)$ is the maximal [square-free](https://en.wikipedia.org/wiki/Square-free_polynomial) divisor of $f(X)$. This relation is implied by three points:
+
+ - $f'(X)$ is the formal derivative of $f(X)$
+ - $g(X)$ is the greatest common divisor of $f(X)$ and $f'(X)$
+ - $h(X) = f(X) / g(X)$.
+
+To see this, let $f(X) = \prod_i f_i(X)^{m_i}$. Then $f'(X) = \sum_i m_i f_i(X)^{m_i - 1} f_i'(X) \prod_{j \neq i} f_j(X)$. And $g(X) = \prod_i f_i(X)^{m_i-1}$ because every factor in the product appears in all terms of $f'(X)$, but there will be one term that limits the multiplicity to $m_i-1$. And from this it follows that $h(X) = f(X) / g(X)$.
+
+Note that is is possible to eliminate $g(X) = f(X) / h(X)$ from the above bullet points. Therefore, two relations remain to be shown: the formal derivative, and the gcd.
+
+### Greatest Common Divisor
+
+Let $a(X)$ and $b(X)$ be randomized Bézout coefficients such that 
+
+$$ a(X) f(X) + b(X) f'(X) = g(X) \enspace .$$
+
+Specifically, the prover computes coefficients $a^\star(X)$ and $b^\star(X)$ using the [extended Euclidean algorithm](https://en.wikipedia.org/wiki/Extended_Euclidean_algorithm). Then he samples a random factor $k \in \mathbb{F}$ and sets $a(X) = a^\star(X) + k f'(X)$ and $b(X) = b^\star(X) - k f(X)$.
+
+The verifier probes this identity of polynomials in a randomly chosen point $z \xleftarrow{\$} \mathbb{F}$.
+
+### Formal Derivative
+
+The formal derivative of a polynomial $f(X) = \sum_{i=0}^d c_i X^i$ is $f'(X) = \sum_{i=1}^d c_i i X^{i-1}$. The correct relation between two polynomial oracles $[f(X)]$ and $[f'(X)]$ can be verified by probing the relation $X \cdot f'(X) = (N_d \circ f)(X)$, where $N_d(X) = \sum_{i=1}^d iX^i$ is the natural arithmetic polynomial up to some degree bound $d \geq \deg(f(X))$, and where $\circ$ denotes the Hadamard (coefficient-wise) product.
+
+Note that the natural arithmetic polynomial can be evaluated efficiently via the following recursive relation, which assumes that $d$ is one less than a power of 2.
+
+$$ N_d(X) = \sum_{i=1}^d i X^i = \sum_{i=0}^{\frac{d+1}{2}-1} i X^i + \sum_{i=0}^{\frac{d+1}{2}-1} (\frac{d+1}{2} + i) X^{\frac{d+1}{2}+i} $$
+
+$$ = \sum_{i=0}^{\frac{d+1}{2} - 1} iX^i + X^{\frac{d+1}{2}} \cdot \left( \frac{d+1}{2} \sum_{i=0}^{\frac{d+1}{2}-1} X^i + \sum_{i=0}^{\frac{d+1}{2}-1}iX^i \right) $$
+
+$$ = \left(1 + X^{\frac{d+1}{2}} \right) \sum_{i=0}^{\frac{d+1}{2}-1} iX^i + \frac{d+1}{2} \cdot X^{\frac{d+1}{2}} \cdot \frac{X^{\frac{d+1}{2}}-1}{X-1}$$
+
+$$ = \left(1 + X^{\frac{d+1}{2}} \right) N_{\frac{d+1}{2}-1}(X) + \frac{d+1}{2} \cdot \frac{X^{d+1}-X^{\frac{d+1}{2}}}{X-1}$$
```

Comment:
Evaluating natural arithmetic polynomials reminds me of an old trick learned in high school, which I think may have a more concise formula (and maybe faster?). The idea is as follows.

To evaluate $\sum_{i=1}^d iX^i$, first denote it by $a$, i.e.,

$$
a = \sum_{i=1}^d i X^i
$$

Then both sides by $X$

$$
aX=\sum_{i=1}^d iX^{i+1} = \sum_{i=2}^{d+1}(i-1)X^i
$$

Then take the difference between the above two equations

$$
a(X-1) = dX^{d+1} - \sum_{i=2}^d X^i - X
$$

Then apply the geometric sum. With further simplifications, it is finally reduced to

$$
a = \frac{(dX-d-1)X^{d+1}+X}{(X-1)^2}
$$

Reply:
Thanks for sharing! I think the trick works and gives rise to a faster evaluation. However, I'm not sure about the last formula ... What I get is this:

$$ X(dX^d - 1) - X^2 \frac{X^{d-1}-1}{X-1} $$

Reply:
I think you forgot to divide it by $X-1$. Apart from this, they are actually the same. I just simplified it such that there is only one exponentiation and one division.

Reply:
Confirmed. Your analysis is entirely correct.